src/olympia/lib/git.py

```diff
@@ -0,0 +1,168 @@
+
+
+class AddonGitRepository(object):
+
+    def __init__(self, addon_id, package_type='package'):
+        assert package_type in ('package', 'source')
+
+        self.git_repository_path = os.path.join(
+            settings.GIT_FILE_STORAGE_PATH,
+            str(addon_id),
+            package_type)
+
+        if not os.path.exists(self.git_repository_path):
+            os.makedirs(self.git_repository_path)
+            self.git_repository = pygit2.init_repository(
+                path=self.git_repository_path,
+                bare=False)
+            # Write first commit to 'master' to act as HEAD
+            tree = self.git_repository.TreeBuilder().write()
+            self.git_repository.create_commit(
+                'HEAD',  # ref
+                self.get_author(),  # author
+                self.get_author(),  # commitor
+                'Initializing repository',  # message
+                tree,  # tree
+                [])  # parents
+
+            log.debug('Initialized git repository {path}'.format(
+                path=self.git_repository_path))
+        else:
+            self.git_repository = pygit2.Repository(self.git_repository_path)
+
+    @classmethod
+    def extract_and_commit_from_file_obj(cls, file_obj, channel):
+        """Extract all files from `file_obj` and comit them.
+
+        This is doing the following:
+
+        * Create a temporary `git worktree`_
+        * Remove all files in that worktree
+        * Extract the zip behind `file_obj` into the worktree
+        * Commit all files
+
+        Kinda like doing...
+
+        * rm -rf worktree/*
+        * unzip file.zip -d worktree/
+        * git commit -a worktree/*
+
+        This ignores the fact that there may be a race-condition of two
+        versions being created at the same time. Since all relevant file based
+        work is done in a temporary worktree there won't be any conflicts and
+        usually the last upload simply wins the race and we're setting the
+        HEAD of the branch (listed/unlisted) to that specific commit.
+
+        .. _`git worktree`: https://git-scm.com/docs/git-worktree
+        """
+        addon = file_obj.version.addon
+        repo = cls(addon.id)
+
+        with TemporaryWorktree(repo.git_repository) as worktree:
+            # Now extract the zip to the workdir
+            zip_file = SafeZip(file_obj.current_file_path, force_fsync=True)
+            zip_file.extract_to_dest(worktree.path)
+
+            # Stage changes
+            worktree.repo.index.add_all()
+            worktree.repo.index.write()
+            tree = worktree.repo.index.write_tree()
+
+            # Create an orphaned commit
+            message = (
+                'Create new version {version} ({version_id}) for '
+                '{addon} from {file_obj}'.format(
+                    version=repr(file_obj.version),
+                    version_id=file_obj.version.id,
+                    addon=repr(addon),
```

Comment:
`repr()` will use the addon name in the current locale - which would be `en-US` in a celery task so okay... but will we ever run this in the request/response?  (if so should we force the locale before doing `repr`?)

Reply:
Forced in 692ef03618